Self Healing Scorecards (#2615)
port-labs/port-docs: docs/guides/all/self-heal-scorecards-with-ai.md

---
displayed_sidebar: null

> description: Learn how to use Port's AI capabilities to automatically detect scorecard degradation and create GitHub issues for Copilot to fix them.

Reviewer: ```suggestion
description: Learn how to use Port's AI capabilities to detect scorecard degradation and automatically use a coding agent like GitHub Copilot to fix it.
```

---

# Self-Heal Scorecards with AI

> Scorecards in Port help you evaluate the maturity, production readiness, and engineering quality of entities in your software catalog. However, when scorecard statistics degrade, manual intervention is often required to identify and fix the issues. This guide shows you how to create an AI-powered system that automatically detects scorecard degradation, creates GitHub issues for remediation, and assigns them to GitHub Copilot for automated code fixes.

Reviewer: the "create github issue" is the means to the end. If we had a direct API to trigger Copilot we would do that.
Let's simplify it and not mention it, like "detects scorecard degradation and trigger Github Copilot for automated code fixes"

<img src="/img/guides/self-healing-scorecard-workflow.jpg" border="1px" width="100%" />

## Common use cases

- **Maintain engineering standards** by detecting missing license files, code owners, or deployment configurations

> - **Track code quality metrics** and generate issues for missing linters, tests, or security scanning

Reviewer: Not "generate issues"

- **Ensure compliance** by monitoring regulatory requirements, security protocols, and data protection measures

## Prerequisites

This guide assumes the following:
- You have a Port account and have completed the [onboarding process](https://docs.port.io/getting-started/overview)
- [Port's GitHub app](https://docs.port.io/build-your-software-catalog/sync-data-to-catalog/git/github/) is installed in your account
- [Port's AI capabilities](https://docs.port.io/ai-agents/overview#getting-started-with-ai-agents) are enabled in your account

> - You have access to GitHub Copilot in your repositories

Reviewer: While this guide described Github Copilot, you can replace it with any other coding agent you have that can be triggered via an API

## Set up data model

We will create and configure blueprints to support our AI-powered self-healing scorecard workflow. This includes updating the GitHub repository blueprint with scorecard-relevant properties and creating new blueprints for tracking AI agent tasks.

> ### Update GitHub repository blueprint

Reviewer: Do we have a guide that already explains how to set up a service scorecard? maybe proudction readiness?
It feels like you must define this scorecard for the guide to work, while in reality you just need to have A scorecard

            "type": "boolean"
          },
          "test_coverage": {
            "title": "Test Coverage %",
            "description": "Percentage of code covered by tests",
            "icon": "Checklist",
            "type": "number"
          },
          "has_code_owners": {
            "title": "Has Code Owners",
            "description": "Whether CODEOWNERS file is configured",
            "icon": "Users",
            "type": "boolean"
          },
          "license": {
            "type": "string",
            "title": "License",
            "format": "markdown"
          },
          "contributing_guide": {
            "type": "string",
            "title": "Contributing Guide",
            "format": "markdown"
          },
          "has_linter": {
            "type": "boolean",
            "title": "Has Linter"
          },
          "manifest": {
            "type": "string",
            "title": "Package JSON"
          }
        },
        "required": []
      },
      "mirrorProperties": {},
      "calculationProperties": {},
      "aggregationProperties": {},
      "relations": {}
    }
    ```
    </details>

5. Click `Save` to update the blueprint.

> ### Create GitHub issue blueprint

Reviewer: As I see we show this part over and over, make sense to have a shared component?

          "many": false
        }
      }
    }
    ```
    </details>

5. Click `Create` to save the blueprint.

### Create AI agent tasks blueprint

This blueprint will track tasks created by the AI agent for scorecard remediation:

1. Go to the [builder](https://app.getport.io/settings/data-model) page of your portal.
2. Click on `+ Blueprint`.
3. Click on the `{...} Edit JSON` button.
4. Copy and paste the following JSON configuration:

    <details>
    <summary><b>AI agent tasks blueprint (Click to expand)</b></summary>

    ```json showLineNumbers
    {
      "identifier": "aiAgentTask",
      "title": "AI Agent Task",
      "icon": "AI",
      "schema": {
        "properties": {
          "title": {
            "title": "Title",
            "type": "string"
          },
          "description": {
            "title": "Description",
            "type": "string",
            "format": "markdown"
          },
          "labels": {
            "title": "Labels",
            "type": "array"
          },
          "source": {
            "title": "Source",
            "type": "string"
          }
        },
        "required": []
      },
      "mirrorProperties": {

>         "pr_status": {

Reviewer: @PeyGis in our environment we also have what we called "work status", which we can rename to "AI agent status". Worth mirroring it as well.
Instead of explaining how to build it, let's have a separate guide (in a different PR) on tracking the AI status of pull requests.

Author: Sorry I don't get this. Can you clarify?

          "target": "githubRepository",
          "required": true,
          "many": false
        }
      }
    }
    ```
    </details>

5. Click `Create` to save the blueprint.

### Update GitHub integration mapping

Update the GitHub integration configuration to link pull requests to AI agent tasks for visibility and tracking:

1. Go to the [data sources](https://app.getport.io/settings/data-sources) page of your portal.
2. Find your GitHub integration and click on it.
3. Go to the `Mapping` tab.
4. Add the following YAML block to map pull requests to AI agent tasks:

    <details>
    <summary><b>Updated GitHub integration configuration (Click to expand)</b></summary>
    
    :::tip JQ Pattern Explanation
    We search for the "task-r" pattern because it is used when saving an AI task. The agent is programmed to include the task identifier in the pull request description. This allows us to map pull requests to their corresponding tasks when fetching them.
    :::

    ```yaml showLineNumbers
    resources:
      - kind: pull-request
        selector:
          query: (.body // "") | test("task-r_[a-zA-Z0-9]{16}")
        port:
          entity:
            mappings:

>               identifier: (.body // "") | match("task-r_[a-zA-Z0-9]{16}").string | tostring

Author: I already have an explanation above it in the guide

              blueprint: '"aiAgentTask"'
              properties: {}
              relations:
                pull_request: .id|tostring
    ```
    </details>

5. Click `Save` to update the integration configuration.

> ## Set up scorecards

Reviewer: And state a pre-requisite is that you have a scorecard define, like production readinesss
